From 9d945df9b8e4dbab356c29c30b240de85f3d012b Mon Sep 17 00:00:00 2001
From: Lans Zhang <jia.zhang@windriver.com>
Date: Fri, 18 Jan 2013 14:59:21 +0800
Subject: [PATCH] x86: mce: Split timer init

commit 26c3c283c5b08dd250279c06ba3ab5b094bbacc3 upstream

Split timer init function into the init and the start part, so the
start part can replace the open coded version in CPU_DOWN_FAILED.

Signed-off-by: Thomas Gleixner <tglx@linutronix.de>
Signed-off-by: Chen Gong <gong.chen@linux.intel.com>
Acked-by: Borislav Petkov <borislav.petkov@amd.com>
Signed-off-by: Tony Luck <tony.luck@intel.com>
[make minor changes to adapt to rt.]
Signed-off-by: Lans Zhang <jia.zhang@windriver.com>

diff --git a/arch/x86/kernel/cpu/mcheck/mce.c b/arch/x86/kernel/cpu/mcheck/mce.c
index 87eb444..d68459a 100644
--- a/arch/x86/kernel/cpu/mcheck/mce.c
+++ b/arch/x86/kernel/cpu/mcheck/mce.c
@@ -1267,7 +1267,7 @@ static unsigned long check_interval = 5 * 60; /* 5 minutes */
 static DEFINE_PER_CPU(unsigned long, mce_next_interval); /* in jiffies */
 static DEFINE_PER_CPU(struct hrtimer, mce_timer);
 
-static enum hrtimer_restart mce_start_timer(struct hrtimer *timer)
+static enum hrtimer_restart mce_timer_fn(struct hrtimer *timer)
 {
 	unsigned long *n;
 
@@ -1634,25 +1634,29 @@ static void __mcheck_cpu_init_vendor(struct cpuinfo_x86 *c)
 	}
 }
 
-static void __mcheck_cpu_init_timer(void)
+static void mce_start_timer(unsigned int cpu, struct hrtimer *t)
 {
-	struct hrtimer *t = &__get_cpu_var(mce_timer);
 	unsigned long *n = &__get_cpu_var(mce_next_interval);
 
-	hrtimer_init(t, CLOCK_MONOTONIC, HRTIMER_MODE_REL);
-	t->function = mce_start_timer;
-
-	if (mce_ignore_ce)
+	if (mce_ignore_ce || !*n)
 		return;
 
 	*n = check_interval * HZ;
-	if (!*n)
-		return;
 
 	hrtimer_start_range_ns(t, ns_to_ktime(jiffies_to_usecs(*n) * 1000ULL),
 			       0 , HRTIMER_MODE_REL_PINNED);
 }
 
+static void __mcheck_cpu_init_timer(void)
+{
+	struct hrtimer *t = &__get_cpu_var(mce_timer);
+	unsigned int cpu = smp_processor_id();
+
+	hrtimer_init(t, CLOCK_MONOTONIC, HRTIMER_MODE_REL);
+	t->function = mce_timer_fn;
+	mce_start_timer(cpu, t);
+}
+
 /* Handle unconfigured int18 (should never happen) */
 static void unexpected_machine_check(struct pt_regs *regs, long error_code)
 {
@@ -2351,6 +2355,7 @@ static int __cpuinit
 mce_cpu_callback(struct notifier_block *nfb, unsigned long action, void *hcpu)
 {
 	unsigned int cpu = (unsigned long)hcpu;
+	struct hrtimer *t = &per_cpu(mce_timer, cpu);
 
 	switch (action) {
 	case CPU_ONLINE:
@@ -2372,6 +2377,7 @@ mce_cpu_callback(struct notifier_block *nfb, unsigned long action, void *hcpu)
 	case CPU_DOWN_FAILED:
 	case CPU_DOWN_FAILED_FROZEN:
 		smp_call_function_single(cpu, mce_reenable_cpu, &action, 1);
+		mce_start_timer(cpu, t);
 		break;
 	case CPU_POST_DEAD:
 		/* intentionally ignoring frozen here */
-- 
1.8.3.1

